ci/get_productversion.sh

```diff
@@ -1,77 +1,190 @@
+EOF
+  exit
+}
+
+bad_usage() {
+    echo "${1}"
+    echo
+    usage
     exit 1
-fi
+}
+
+parse_params() {
+    while :; do
+        case "${1-}" in
+            -h | --help) usage ;;
+            -?*)
+                bad_usage "Unknown option: ${1}" ;;
+            *) break ;;
+        esac
+        shift
+    done
+
+    args=("$@")
+    [[ ${#args[@]} -eq 0 ]] && usage && exit 1
+    [[ ${#args[@]} -ne 1 ]] && bad_usage "Too many arguments"
+
+    return 0
+}
+
+parse_version_tag() {
+    version_tag="${VERSION_TAG}"
+    if [ -z "${version_tag}" ]; then
+        version_tag=$(git tag -l --sort -version:refname | head -n 1)
+    fi
+
+    version_regex="^v([0-9]+).([0-9]+).([0-9]+)((-(alpha|beta|rc|sumo)[-.]([0-9]+))(-(alpha|beta|rc).([0-9])+)?)?$"
 
-if [[ $ot_channel == "sumo" ]]; then
-    if [[ $sumo_channel != "" ]]; then
-        internal_version="${sumo_channel_version}"
+    if [[ $version_tag =~ $version_regex ]]; then
+        major_version="${BASH_REMATCH[1]}"
+        minor_version="${BASH_REMATCH[2]}"
+        patch_version="${BASH_REMATCH[3]}"
+        ot_channel="${BASH_REMATCH[6]}"
+        ot_channel_version="${BASH_REMATCH[7]}"
+        sumo_channel="${BASH_REMATCH[9]}"
+        sumo_channel_version="${BASH_REMATCH[10]}"
     else
-        internal_version="${ot_channel_version}"
+        echo "Error: Tag does not match required version regex: ${version_tag}" >&2
+        exit 1
+    fi
+
+    if [[ $ot_channel == "sumo" ]]; then
+        if [[ $sumo_channel != "" ]]; then
+            build_version="${sumo_channel_version}"
+        else
+            build_version="${ot_channel_version}"
+        fi
+    elif [[ $ot_channel != "" ]]; then
+        build_version="${ot_channel_version}"
+    fi
+
+    if [[ $OVERRIDE_BUILD_VERSION != "" ]]; then
+        number_regex='^[0-9]+$'
+        if ! [[ $OVERRIDE_BUILD_VERSION =~ $number_regex ]]; then
+            echo "Error: OVERRIDE_BUILD_VERSION is not a number" >&2
+            exit 1
+        fi
+
+        build_version="${OVERRIDE_BUILD_VERSION}"
     fi
-elif [[ $ot_channel != "" ]]; then
-    internal_version="${ot_channel_version}"
-fi
+}
 
-if [[ $OVERRIDE_INTERNAL_VERSION != "" ]]; then
-    number_regex='^[0-9]+$'
-    if ! [[ $OVERRIDE_INTERNAL_VERSION =~ $number_regex ]]; then
-        echo "Error: OVERRIDE_INTERNAL_VERSION is not a number" >&2
+# Validate version information using the Windows ProductVersion requirements.
+validate() {
+    if [[ -z "${major_version}" ]]; then
+        echo "Major version cannot be empty"
         exit 1
     fi
 
-    internal_version="${OVERRIDE_INTERNAL_VERSION}"
-fi
+    if [[ $major_version -lt 0 ]]; then
+        echo "Major version cannot be less than 0"
+        exit 1
+    fi
 
-# Validation
-if [[ $major_version -gt 255 ]]; then
-    echo "Major version cannot be greater than 255"
-    exit 1
-fi
+    if [[ $major_version -gt 255 ]]; then
+        echo "Major version cannot be greater than 255"
```

Comment:
Where does this requirement come from? Windows?

Reply:
I linked to the documentation in a comment further down: https://learn.microsoft.com/en-us/windows/win32/msi/productversion.